signer/tests/integration/transaction_signer.rs

```diff
@@ -0,0 +1,72 @@
+use signer::storage;
+use signer::testing;
+
+use futures::StreamExt;
+use rand::RngCore;
+use sqlx::Executor;
+
+async fn test_environment(
+    pool: sqlx::PgPool,
+) -> testing::transaction_signer::TestEnvironment<impl FnMut() -> storage::postgres::PgStore> {
+    let num_signers = 3;
+    let context_window = 3;
+    let test_databases: Vec<_> = futures::stream::iter(0..num_signers)
+        .then(|_| async { new_database(&pool).await })
+        .collect()
+        .await;
+
+    let mut idx = 0;
+
+    testing::transaction_signer::TestEnvironment {
+        storage_constructor: move || {
+            idx = (idx + 1) % test_databases.len();
+            storage::postgres::PgStore::from(test_databases.get(idx).unwrap().clone())
+        },
+        context_window,
+        num_signers,
+    }
+}
+
+static MIGRATOR: sqlx::migrate::Migrator = sqlx::migrate!();
+
+async fn new_database(pool: &sqlx::PgPool) -> sqlx::PgPool {
+    let mut rng = rand::rngs::OsRng;
+    let db_name = format!("test_db_{}", rng.next_u64());
+
+    let create_db = format!("CREATE DATABASE \"{}\";", db_name);
+    pool.execute(create_db.as_str()).await.expect("Nooo");
+
+    let base_url =
+        std::env::var("DATABASE_URL").expect("DATABASE_URL must be set in the environment");
```

Comment:
Theoretically someone could have different credentials to their database though so I think this is a bit more flexible w.r.t the developers local environment. In practice I assume we're all running with the same configuration using docker compose - so I'd also be fine with hard-coding it. Though I don't see the benefit right now, why is that safer?

Reply:
> Though I don't see the benefit right now, why is that safer?

I'm always a little worried that I'd have production environment variables set in my shell, forget, and then do something like `cargo test`. But yeah, this is only an issue if we use the same environment variable name for the database url for tests and for prod, and when we are actually running a live signer.

Reply:
Ah good point. In this case I think we're good since we'll 1. not run a live signer and 2. the production environment is likely to be managed fairly isolated from our local environments (using e.g. helm charts and k8s deployments) so I don't think we'll ever have a `DATABASE_URL` pointing to a production database locally - instead we'll probably use `helm install` or `kubectl apply` to configure any live environments.

Reply:
And worst case scenario is that our tests would put some load on the production DB. It would still be isolated and not corrupt any real data 😅